library/rmc/src/lib.rs

```diff
@@ -51,3 +53,82 @@ pub fn nondet<T>() -> T {
 #[inline(never)]
 #[rustc_diagnostic_item = "RmcExpectFail"]
 pub fn expect_fail(_cond: bool, _message: &str) {}
+
+/// Given an array `arr` of length `LENGTH`, this function returns a **valid**
+/// slice of `arr` with non-deterministic start and end points.  This is useful
+/// in situations where one wants to verify that all possible slices of a given
+/// array satisfy some property.
+///
+/// # Example:
+///
+/// ```rust
+/// let arr = [1, 2, 3];
+/// let slice = rmc::nondet_slice(&arr);
+/// foo(slice); // where foo is a function that takes a slice and verifies a property about it
+/// ```
+pub fn nondet_slice<T, const LENGTH: usize>(arr: &[T; LENGTH]) -> &[T] {
+    let (from, to) = nondet_range::<LENGTH>();
+    &arr[from..to]
+}
+
+/// A mutable version of the previous function
+pub fn nondet_slice_mut<T, const LENGTH: usize>(arr: &mut [T; LENGTH]) -> &mut [T] {
+    let (from, to) = nondet_range::<LENGTH>();
+    &mut arr[from..to]
+}
+
+fn nondet_range<const LENGTH: usize>() -> (usize, usize) {
+    let from: usize = nondet();
+    let to: usize = nondet();
+    assume(to <= LENGTH);
+    assume(from <= to);
+    (from, to)
+}
+
+/// A struct that creates a slice of type `T` with a non-deterministic length
+/// between `0..MAX_SLICE_LENGTH` and with non-deterministic content.  This is
+/// useful in situations where one wants to verify that all slices with any
+/// content and with a length up to `MAX_SLICE_LENGTH` satisfy a certain
+/// property
+///
+/// # Example:
+///
+/// ```rust
+/// let slice = rmc::NonDetSlice::<u8, 5>::new();
+/// foo(&slice); // where foo is a function that takes a slice and verifies a property about it
+/// ```
+pub struct NonDetSlice<T, const MAX_SLICE_LENGTH: usize> {
+    arr: [T; MAX_SLICE_LENGTH],
+    slice_len: usize,
+}
+
+impl<T, const MAX_SLICE_LENGTH: usize> NonDetSlice<T, MAX_SLICE_LENGTH> {
+    pub fn new() -> Self {
```

Comment:
Instead of new, I would recommend calling this from a function like the others in this module:

```rust
pub fn nondet_slice<T, const N: usize>() -> NonDetSlice<T, N>
```

Or we can just generate an array here too.
```rust
pub fn nondet_array<T, const N: usize>() -> [T; N]
```

Reply:
> Instead of new, I would recommend calling this from a function 

Good idea! I will add a function.

> Or we can just generate an array here too

Generating an array is already possible with `nondet`:
```rust
let arr: [u8; 5] = rmc::nondet();
```